### ISSUE TITLE
[BUG]: Potential Data Leakage Leading to High Accuracy in PyCaret Regression Models

### ISSUE THREAD

**Issue by Nozawa-610** [pycaret/pycaret#3829]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [ ] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

I am encountering an issue with unusually high accuracy scores in regression models when using PyCaret, specifically with tree-based models such as Extra Trees and Random Forest. This issue was first noticed in PyCaret version 2.3.10 and continues to occur in version 3.2.0.

The problem arises during the testing phase, where the performance metrics such as accuracy, kappa, and MCC are showing perfect scores, which is highly unusual and indicative of potential data leakage.

This anomaly in performance metrics is a significant concern as it suggests a fundamental issue in model evaluation or data handling within PyCaret, especially considering that these results were observed across different models and setups. I am looking for insights or suggestions to understand and resolve this issue while utilizing the full capabilities of PyCaret.

### Reproducible Example

```python
from pycaret.regression import *

# Prepare a sample dataset (emulating the actual data structure)
# df = ... (example of dataset)
numeric_features = ['feature1', 'feature2', ...]
categorical_features = ['feature3', 'feature4', ...]

# Setting up PyCaret
exp1 = setup(data=df, target='target', train_size=0.8, numeric_features=numeric_features, categorical_features=categorical_features, session_id=1234, normalize=True, normalize_method='minmax', feature_selection=True, feature_selection_method='classic', n_features_to_select=0.8, imputation_type="iterative", n_jobs=-1, use_gpu=True)

# Creating Extra Trees model
et = create_model('et')

# Finalizing the model
final_model = finalize_model(et)

# Predicting on test data
result = predict_model(final_model, data=get_config('X_test'))

# Getting true values
y_true = get_config('y_test')

# Combining results
result_file = pd.concat([y_true, result['Label']], axis=1)

# Custom performance evaluation function
# accuracy, kappa, mcc = evaluate_model_performance(result_file)
# evaluate_model_performance() is a custom function that calculates accuracy, weighted kappa, and MCC, by rounding off the predicted values as the true values are discrete.
```

### Expected Behavior

The expected behavior is to obtain realistic performance metrics (Accuracy, Kappa, MCC) from appropriate predictions on the test data.

### Actual Results

```python-traceback
The actual results showed anomalously high performance metrics when predictions were made on the test data. Specifically, the following results were observed:

Accuracy: 1.0
Kappa: 1.0
MCC: 1.0
These results suggest the possibility of data leakage or other unknown issues.
```

### Installed Versions

<details>
- PyCaret Version: 2.3.10, 3.2.0
- Python Version: 3.7.13, 3.10.13
- Operating System: Windows 10
</details>

**Comment by celestinoxp** [comment]
@ngupta23 I believe there may be an overfitting problem, you can take a look here?

**Comment by Nozawa-610** [comment]
Thanks for your comments and suggestions. 

I believe get_config('X_test') is a test dataset automatically allocated by PyCaret's setup() function and is not used in the model training or cross-validation process.

**Comment by SarthakNikhal** [comment]
Is it still an overfitting issue or is it a pycaret issue?

**Comment by celestinoxp** [comment]
The dataset data is divided into 2 parts, in this case, 80% for training and 20% for testing.
In this case, the user is using this data used internally by Pycaret to make variations, so as the algorithm knows this data, it gives 100% accuracy...

If in predict_model() you use unknown data you will see that the accuracy value is different...

In the next line, you should not use data=get_config('X_test')
result = predict_model(model_final, data=get_config('X_test'))

instead, it should use unknown data to make a prediction, for example:
result = predict_model(final_model, data=unseen_data)

**Comment by Yard1** [comment]
Hi @Nozawa-610, once you call `finalize_model`, the model will be retrained on the entire dataset, including the test set. From thereon, you should be using it on new data only. If you ommit the `finalize_model` from your script, you will see that the scores differ, as the test set won't be used for model training.

**Comment by Nozawa-610** [comment]
Hi @Yard1,

Thank you so much for your insightful comment. It was incredibly helpful. Your explanation about the behavior of finalize_model in PyCaret clarified the issue I was experiencing. Upon omitting finalize_model from my script, I observed the expected variation in scores, confirming that the test set was indeed not being used in model training.

Your guidance not only resolved my immediate issue but also provided me with a better understanding of how to correctly use PyCaret in future projects. I am grateful for your assistance and the time you took to address my concern.

Best regards,
Nozawa-610
